xarray/core/duck_array_ops.py

```diff
@@ -161,7 +161,7 @@ def trapz(y, x, axis):
 def asarray(data):
     return (
         data
-        if (isinstance(data, dask_array_type) or hasattr(data, "__array_function__"))
+        if (is_duck_dask_array(data) or hasattr(data, "__array_function__"))
```

Comment:
I *think* this is right

Reply:
I think we should define another helper function, `is_duck_array()`, to use in situations like this:
```python
def is_duck_array(data):
    # no need to check for dask explicitly now that it implements __array_function__
    # depending on backwards compat, potentially we still need to explicitly include
    # np.ndarray here?
    return hasattr(data, "__array_function__")
```

The reason for `is_duck_array()` rather than `is_array_like()` is that a "duck array" is a strong condition than "array like", at least as [is commonly used by NumPy](https://numpy.org/neps/nep-0022-ndarray-duck-typing-overview.html):

- "Array like"  means "convertible to a NumPy array", i.e., with `np.asarray()`.
- "Duck array" means "can be computed on like a NumPy array".

Examples:

- Python lists, zarr.Array and netCDF4.Variable are "array like" but not "duck arrays"
- CuPy and sparse arrays are "duck arrays" but not "array like" (`np.asarray` fails)
- NumPy, dask and pint arrays are both "array like" _and_ "duck arrays"

It is valid to pass either a "duck array" or an "array like" into an xarray data structure, but `.data` should always be a duck array.

(`._data` can sometimes include "arrays likes", especially xarray's own internal "lazy arrays" that support lazy indexing. This is worth cleaning up eventually, but at least it isn't exposed to users who use `.data`.)

Reply:
@rpmaner can you make this change too please?

Reply:
@dcherian Yes, I can do this as well. Should *all* `hasattr(data, "__array_function__")` checks be replaced with the above `is_duck_array`, or just in the places discussed in this PR? And should all `is_duck_dask_array` checks also be replaced with `is_duck_array` (assuming dask has implemented `__array_function__`)?

Reply:
just to make sure, I think we should rename `is_array_like` to `is_duck_array`, and also have it check for `__array_ufunc__` and `__array_function__`. Not sure if I'm missing anything, though.

`is_duck_dask_array` should be fine as is?

Reply:
> just to make sure, I think we should rename `is_array_like` to `is_duck_array`, and also have it check for `__array_ufunc__` and `__array_function__`. Not sure if I'm missing anything, though.

Makes sense to me.

> `is_duck_dask_array` should be fine as is?

You're right, `is_duck_dask_array` should be okay since it checks if the item is a dask collection.